source/blog/2017-10-10-glimmer-progress-report.md

```diff
@@ -0,0 +1,645 @@
+```
+
+But what happens if I want to set the width of the underlying `img` element via
+its `width` attribute?
+
+```jsx
+<HiResImage width="100%" src="corgi.jpg" hiResSrc="corgi@2x.jpg" />
+```
+
+This won't work because the only attributes or properties that get set are the
+ones we've manually listed in our `render()` method! The `width` attribute here
+will just get ignored.
+
+We have a few options, but none of them are that great.
+
+* We could enumerate all of the possible valid `img` attributes that might get
+  passed in, but that is error-prone (new attributes get added all the time) and
+  takes a lot of code.
+* We could use the spread operator (`...props`), but that will set *everything*
+  passed in as an attribute, even "known" props that aren't attributes, like
+  `hiResSrc`.
+* If we're using Babel, we can use rest syntax in object destructuring to
+  separate "known" and "unknown" props: `let { src, hiResSrc, ...attrs } =
+  props`. But this requires non-trivial runtime work, and means any props passed
+  in by accident will now be treated as an attribute.
+* In Preact, the problem is even trickier because it will always set the `width`
+  property no matter what, never the attribute. And setting the `width`
+  *property* to `"100%"` results in an image zero pixels wide.
+
+With Glimmer.js, you explicitly disambiguate between properties and attributes via
+the presence of the `@` sigil. In symmetry with HTML, attributes do not have `@`,
+while component arguments (`props` in React parlance) do:
+
+```hbs
+<HiResImage @src="corgi.jpg" @hiResSrc="corgi@2x.jpg" width="100%" />
+```
+
+In this example, `src` and `hiResSrc` are JavaScript values passed as arguments
+to the component object, and `width` is serialized to a string and set as an
+attribute.
+
+"But wait," you ask, "if components aren't required to have a single root
+element anymore, where do attributes go?"
+
+With recent changes in Glimmer VM, we can now support an `...attributes` syntax
+that we've colloquially started calling "splattributes" (because they "splat"
+attributes from the outside onto an element).
+
+In our case, the Glimmer.js version of the `HiResImage` component might look like this:
+
+```js
+import Component, { tracked } from '@glimmer/component';
+
+export default class HiResImage extends Component {
+  @tracked hiRes = false;
+
+  showHiRes() {
+    this.hiRes = true;
+  }
+}
+```
+
+```hbs
+{{#if hiRes}}
+  <img src={{@hiResSrc}} ...attributes>
+{{else}}
+  <img src={{@src}} {{on click=(action showHiRes)}} ...attributes>
+{{/if}}
+```
+
+Here, any attributes passed in on the invoking side will be "splatted" onto the
+appropriate element.
+
+So what happens if you try to pass attributes to a component that doesn't have
+`...attributes`? At runtime, you'll get a hard error telling you that the
+component should add `...attributes` to one or more elements. We can probably
+produce compile-time errors in the majority of less dynamic cases, too.
```

Comment:
"less dynamic cases as well".

Or, "Additionally, web can probably..." without ", too."

", too." just reads a bit odd to me.

Reply:
I would say "and we can in the future look at producing compile-time errors" and leave it at that.